### DISCUSSION TITLE
Need to include WebView2 runtime in compiled windows executable

### DISCUSSION THREAD

**Discussion by HGS-ryanbloom** [tauri-apps/tauri#4774] [Q&A] [answered]
Hello, I need to be able to ship a single standalone executable file, but there are instances where a user might not have WebView2 installed on their windows system (because it might be too old, too bare, etc). I'm unable to use the installer which wix provides, and must instead use just the executable. Is there any way where I can bundle the WebView2 Runtime in the executable? Would it work as a sidecar? [Here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/static) there is an article on distributing a WebView2 app as a single binary but I'm wondering if this can be done with a tauri app as well. Thanks for any help!

**Comment by FabianLars** [comment, marked as answer]
> Is there any way where I can bundle the WebView2 Runtime in the executable?

No, you can't bundle the whole runtime inside the _executable_. You can only bundle it alongside your binary, like in the .msi file the bundler creates, or a .zip file if you wanna create your own stuff.

> [Here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/static) there is an article on distributing a WebView2 app as a single binary

This is only about the bindings to the webview, not the runtime itself. We already do what the article described automatically, otherwise you'd have to bundle a .dll file alongside your binary.

So since you said you need it to be a single executable file, there's only really one way to go here. You need to install the webview from within your executable.
A tldr of that:
In your main.rs file before the tauri builder you need to check if it's installed already by searching for registry keys (they are listed in our main.wxs file for example), i'd suggest using the winreg crate for that, which would look like this
```rs
if RegKey::predef(HKEY_LOCAL_MACHINE).open_subkey("SOFTWARE\\WOW6432Node\\Microsoft\\EdgeUpdate\\Clients\\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}").is_err() // System-wide 64bit
      && RegKey::predef(HKEY_LOCAL_MACHINE).open_subkey("SOFTWARE\\Microsoft\\EdgeUpdate\\Clients\\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}").is_err() // System-wide 32bit
      && RegKey::predef(HKEY_CURRENT_USER).open_subkey("SOFTWARE\\Microsoft\\EdgeUpdate\\Clients\\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}").is_err() // User-wide 64bit&32bit
      {}
```
In case it's not installed (inside the if body), you download the installer via something like reqwest (if you need windows 7/8 support) or powershell, into a temp dir and execute that via a powershell [Command](https://doc.rust-lang.org/std/process/struct.Command.html).
Alternatively you can embed the _installer_ in the binary and copy that into a temp dir. But again, you can't bundle the runtime itself _inside_ the app.

**Reply by HGS-ryanbloom** [reply]
Thank you for the thorough description. 

I did try bundling the fixed version of WV2 but I had trouble getting Tauri to recognize this installation and I wonder if I would have this same issue with automatically running the installer?

`thread 'main' panicked at 'error while running tauri application: Runtime(CreateWebview(WebView2Error(WindowsError(Error { code: 0x80070002, message: The system cannot find the file specified.
 }))))', src\main.rs:212:10
`

This is the error I got, which never came up when just using the built-in WV2 install on my OS, but instead only during testing when using the installer with a fixed WV2 version.

Essentially, is there any way that I can be sure the Tauri app is going to point to the correct WV2 install location if installed by the exe?

**Reply by FabianLars** [reply]
> Essentially, is there any way that I can be sure the Tauri app is going to point to the correct WV2 install location if installed by the exe?

Not really, it will always just panic, but for what it's worth, I don't know of a single time where it wasn't able to pick up an actually installed WebView2 (via the official installers)

**Reply by HGS-ryanbloom** [reply]
> Alternatively you can embed the installer in the binary and copy that into a temp dir. But again, you can't bundle the runtime itself inside the app.

Could you elaborate a little bit on embedding the installer into the binary? I presume this would be done with a sidecar somehow but I need it to only embed as such on windows.

Sorry for the trouble, and thanks very much again for your responses.

**Reply by FabianLars** [reply]
If you mean tauri's sidecar mechanism, this only works with the .msi bundle.
What i have in mind is this https://doc.rust-lang.org/std/macro.include_bytes.html and then writing these bytes to a file at runtime.

> I need it to only embed as such on windows.

The solution is pretty simple, just annotate everything that's windows-only with `#[cfg(windows)]`

**Reply by HGS-ryanbloom** [reply]
Got it, this is great, thanks again.
